### PR TITLE
Update node-uwp-wrapper for next release

### PR DESCRIPTION
**Pull request by munyirik**
This change includes:
1. Updating referenced Windows and IoT SDK versions.
2. Copies app files to local storage folder when started to give users write access to the root of the app.

### CODE REVIEW

**Comment by munyirik** [comment]
@alanch-ms @msalehmsft could you CR? I need 2 sign-offs. Mahmoud, I'm using process launcher in this change.

File: nodeuwp/StartupTask.h [hunk 1]
```diff
@@ -33,5 +33,9 @@ namespace nodeuwp
 	{
 	public:
 		virtual void Run(Windows::ApplicationModel::Background::IBackgroundTaskInstance^ taskInstance);
+	private: 
+		void CopyFolderSync(Windows::Storage::StorageFolder^ source, Windows::Storage::StorageFolder^ destination);
+		void PopulateArgsVector(std::vector<std::shared_ptr<char>> &argVector, Windows::Data::Xml::Dom::XmlNodeList^ argNodes, bool isStartupScript = false);
+		std::shared_ptr<char> StartupTask::PlatformStringToChar(const wchar_t* str, int strSize);
```
Comment (alanch-ms) [inline nodeuwp/StartupTask.h]: Nits: I don't think you need the 'StartupTask::' qualifer

File: nodeuwp/StartupTask.cpp [hunk 2]
```diff
@@ -141,7 +169,7 @@ void StartupTask::Run(IBackgroundTaskInstance^ taskInstance)
 			PopulateArgsVector(argumentVector, argumentNodes);
 
 			argumentNodes = startupInfoXml->SelectNodes(L"StartupInfo/Script");
-			PopulateArgsVector(argumentVector, argumentNodes);
+			PopulateArgsVector(argumentVector, argumentNodes, true);
 
 			argumentNodes = startupInfoXml->SelectNodes(L"StartupInfo/ScriptArgs");
 			PopulateArgsVector(argumentVector, argumentNodes);
```
Comment (alanch-ms) [inline nodeuwp/StartupTask.cpp]: Any chance of the 'ScriptArgs' actually refers to the a file/script (ie. a relative path)? In that case, Node probably would use the current directory (probably pointing to the install path) and combine it with the relative path.  Thus, the final result would be in the install path.

File: nodeuwp/StartupTask.cpp [hunk 3]
```diff
@@ -120,6 +142,12 @@ void PopulateArgsVector(std::vector<std::shared_ptr<char>> &argVector, XmlNodeLi
 
 void StartupTask::Run(IBackgroundTaskInstance^ taskInstance)
 {
+	StorageFolder^ appFolder = Windows::ApplicationModel::Package::Current->InstalledLocation;
+	StorageFolder^ localFolder = ApplicationData::Current->LocalFolder;
+	// Copy files to this applications local storage so that node can read/write to files
+	// or folders relative to the location of the starup JavaScript file
+	CopyFolderSync(appFolder, localFolder);
```
Comment (alanch-ms) [inline nodeuwp/StartupTask.cpp]: Do you want to copy everything (including the binaries) or just the scripts?

File: nodeuwp/StartupTask.cpp [hunk 4]
```diff
@@ -79,7 +82,18 @@ std::shared_ptr<char> PlatformStringToChar(const wchar_t* str, int strSize)
 	return buffer;
 }
 
-void PopulateArgsVector(std::vector<std::shared_ptr<char>> &argVector, XmlNodeList^ argNodes)
+void StartupTask::CopyFolderSync(StorageFolder^ source, StorageFolder^ destination)
+{
+	// Using xcopy for best performance
+	String^ args = source->Path + "\\*" + " " + destination->Path + " " + "/s /d /y";
+	ProcessLauncherResult^ result = create_task(ProcessLauncher::RunToCompletionAsync("c:\\windows\\system32\\xcopy.exe", args)).get();
```
Comment (alanch-ms) [inline nodeuwp/StartupTask.cpp]: Any chance of 'windows' not located on c:\?

File: nodeuwp/StartupTask.cpp [hunk 5]
```diff
@@ -47,7 +50,7 @@ using namespace Platform::Collections;
 // output to a file (nodeuwp.log) in this applications local storage folder.
 bool useLogger = false;
 
-std::shared_ptr<char> PlatformStringToChar(const wchar_t* str, int strSize)
+std::shared_ptr<char> StartupTask::PlatformStringToChar(const wchar_t* str, int strSize)
```
Comment (alanch-ms) [inline nodeuwp/StartupTask.cpp]: Why std::shared_ptr<char> instead of a std::string?

File: nodeuwp/StartupTask.h (hunk 1 above)
Comment (munyirik) [inline nodeuwp/StartupTask.h]: yeah, I'll take it out.

File: nodeuwp/StartupTask.cpp (hunk 2 above)
Comment (munyirik) [inline nodeuwp/StartupTask.cpp]: The path will be set by Visual Studio (user has no control other than setting it through VS UI) and it's not a relative path.

File: nodeuwp/StartupTask.cpp (hunk 3 above)
Comment (munyirik) [inline nodeuwp/StartupTask.cpp]: The node binaries are very small and few that it wouldn't affect overall performance. I also wanted to keep the code clean and not maintain a list of just a few files to filter.

File: nodeuwp/StartupTask.cpp (hunk 4 above)
Comment (munyirik) [inline nodeuwp/StartupTask.cpp]: The way the images are set up at least for TH2, it's always going to be in c:\.

File: nodeuwp/StartupTask.cpp (hunk 5 above)
Comment (munyirik) [inline nodeuwp/StartupTask.cpp]: Hmm, don't remember why I decided to use a smart pointer. I was thinking maybe it made the conversion simpler but looking at the code I think std::string.c_str() works as well.
Any concerns here or were you just curious?

**Comment by munyirik** [comment]
@msalehmsft @alanch-ms I ended up using filesystem as Artur suggested and it's just as fast as xcopy when I use its directory iterator.

File: nodeuwp/StartupTask.cpp (hunk 2 above)
Comment (alanch-ms) [inline nodeuwp/StartupTask.cpp]: Did you end up changing the current working directory to the local storage folder?  I don't see that change.

Comment (munyirik) [inline nodeuwp/StartupTask.cpp]: I made that change with https://github.com/Microsoft/node-msft/pull/130. It'll only happen before a read or write occurs.

File: nodeuwp/StartupTask.cpp (hunk 5 above)
Comment (alanch-ms) [inline nodeuwp/StartupTask.cpp]: Just curious.  std::string makes it easier if you ever want to find the length or modify the string, etc.

File: nodeuwp/StartupTask.cpp [hunk 6]
```diff
@@ -113,13 +119,40 @@ void PopulateArgsVector(std::vector<std::shared_ptr<char>> &argVector, XmlNodeLi
...
+void StartupTask::CopyFolderSync(StorageFolder^ source, StorageFolder^ destination)
+{
+	wstring from = source->Path->Data();
+	wstring to = destination->Path->Data();
+	copy_options opts = copy_options::recursive | copy_options::update_existing;
+
+	for (directory_iterator next(from), end; next != end; ++next)
+	{
+		copy(next->path(), to / next->path().filename(), opts);
+	}
```
Comment (alanch-ms) [inline nodeuwp/StartupTask.cpp]: (I haven't used directory_iterator's before, so I might be wrong.)  It seems you are iterating (non-recursively) over the items in the top level directory (eg. install dir) and copy recursively to the storage dir.  If "copy()" actually works recursively, why can't you just call copy(from_path, to_path)?

Comment (munyirik) [inline nodeuwp/StartupTask.cpp]: good catch. copy does work recursively so all I need is the copy call with the recursive option.

**Comment by alanch-ms** [comment]
:+1:

**Pull request merged**